rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,775 @@
+        config: A model config used for correctly parametrising the `ConcatenateSparse
+            DenseFeatures` and `ConcatenateSequenceSentenceFeatures` layers.
+
+    Input shape:
+        Tuple of four tensor inputs:
+            sequence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, max_seq_length, input_dim)` where `input_dim` can be
+                different for sparse vs dense tensors.
+            sentence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, 1, input_dim)` where `input_dim` can be different for
+                sparse vs dense tensors, and can differ from that in `sequence_features`.
+            mask_sequence: dense 3-D tensor with the shape `(batch_size, max_sequence_
+                length, 1)`.
+            mask_combined_sequence_sentence: dense 3-D tensor with the shape `(batch_size,
+                max_sequence_length+1, 1)`, i.e. with the 2nd dimension combining the
+                lengths of sequence- and sentence-level features.
+
+    Output shape:
+        3-D tensor with shape: `(batch_size, sequence_length, units)` where `units` is 
+        completely  determined by the internally applied `ConcatenateSparseDenseFeatures`
+        layer and `sequence_length` is completely determined by the internally applied 
+        `ConcatenateSequenceSentenceFeatures` layer.
+
+    Raises:
+        A ValueError if no feature signatures are provided.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        attribute_signature: Dict[Text, List[FeatureSignature]],
+        config: Dict[Text, Any],
+    ) -> None:
+        if not attribute_signature or not (
+            len(attribute_signature.get(SENTENCE, [])) > 0
+            or len(attribute_signature.get(SEQUENCE, [])) > 0
+        ):
+            raise ValueError("The data signature must contain some features.")
+
+        super().__init__(name=f"rasa_input_layer_{attribute}")
+
+        self._tf_layers = {}
+
+        # 1. prepare layers for combining sparse and dense features
+        for feature_type in attribute_signature:
+            if len(attribute_signature[feature_type]) == 0 or feature_type not in [
+                SENTENCE,
+                SEQUENCE,
+            ]:
+                continue
+            self._tf_layers[
+                f"sparse_dense.{feature_type}"
+            ] = ConcatenateSparseDenseFeatures(
+                attribute=attribute,
+                feature_type=feature_type,
+                attribute_signature=attribute_signature[feature_type],
+                config=config,
+            )
+
+        # 2. prepare layer for combining sequence- and sentence-level features
+        self.have_all_feature_types = all(
+            [
+                len(attribute_signature.get(feature_type, [])) > 0
+                for feature_type in [SEQUENCE, SENTENCE]
+            ]
+        )
```

Comment:
can we make it explicit?
```suggestion
        self.have_all_feature_types = SEQUENCE in attribute_signature and SENTENCE in attribute_signature
```

Reply:
I got rid of this part of the code altogether as I'm no longer conditioning anything on what feature types are present (the seq/sent concatenating layer can deal with all of these scenarios).